src/cargo/ops/cargo_new.rs

```diff
@@ -46,10 +49,27 @@ pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
         return Err(human(format!("Destination `{}` already exists",
                                  path.display())))
     }
-    let name = try!(path.file_name().and_then(|s| s.to_str()).chain_error(|| {
-        human(&format!("cannot create a project with a non-unicode name: {:?}",
-                       path.file_name().unwrap()))
-    }));
+    let name = match opts.name {
+        Some(name) => name,
+        None => {
+            let dir_name = try!(path.file_name().and_then(|s| s.to_str()).chain_error(|| {
+                human(&format!("cannot create a project with a non-unicode name: {:?}",
+                               path.file_name().unwrap()))
+            }));
+            if opts.bin {
+                dir_name
+            } else {
+                let new_name = strip_rust_affixes(dir_name);
+                if new_name != dir_name {
+                    let message = format!(
+                        "Note: package will be named `{}`; use --name to override",
```

Comment:
The rest of the file seems to be inconsistent on the casing, so I wasn't sure about that.

I'll change it anyway though.